Polymorphic transforms

qnkhuat opened the pull request:
Introduce transforms to use a polymorphic multimethod-based that unifies the handling of different transform types (MBQL/query transforms and Python transforms) under a common interface.

Motivations:
- Unified Interface - we used to have lots of  conditional logic checking python-transform? scattered throughout to decide between transforms-python.execute/execute-python-transform! and transforms.execute/run-mbql-transform!. 
- Extensibility - we might have more transforms type in the future (llm runner? 🤞 )
- Reduced Coupling - The API layer (transforms.api) no longer needs to know about specific transform implementations.

qnkhuat committed:
make transform polymorphic

qnkhuat committed:
transform deps as well

qnkhuat committed:
test fixes

qnkhuat committed:
fix bad references

qnkhuat committed:
needs to learn the abc

qnkhuat committed:
needs to load the impl

wotbrew commented on enterprise/backend/src/metabase_enterprise/transforms/api.clj:
```diff
@@ -243,13 +241,8 @@
-      (u.jvm/in-virtual-thread*
-       (transforms.execute/run-mbql-transform! transform {:start-promise start-promise
-                                                          :run-method :manual})))
+    (transforms.i/execute! transform {:start-promise start-promise
```
I would personally keep the thread on the outside, its independent from the implementation afaict.

qnkhuat commented on enterprise/backend/src/metabase_enterprise/transforms/interface.clj:
```diff
@@ -0,0 +1,31 @@
+    (transform->transform-type transform)))
+
+(defmulti cancel!
+  "Execute a transform with an option."
```
oh sorry, I meant to delete this as i realized cancel is only for transforms rn.

qnkhuat committed:
remove cancel and bring back virtual thread

wotbrew commented on enterprise/backend/src/metabase_enterprise/transforms/interface.clj:
```diff
@@ -0,0 +1,31 @@
+(defmulti cancel!
+  "Execute a transform with an option."
+  {:added "0.47.0" :arglists '([transform options])}
+  (fn [transform _options]
```
I'd probably have the interface target the 'run' rather than the transform at this level. The fact the http api takes a transform id I would hope changes as it introduces a kind of essential race condition. 

What about `(fn [transform run-id options])`?

wotbrew commented on enterprise/backend/src/metabase_enterprise/transforms/interface.clj:
```diff
@@ -0,0 +1,31 @@
+  (-> transform :source :type keyword))
+
+(defmulti source-db-id
+  "Return the source db id for a given transform."
```
Non blocking, we can change it later - but what about pythonic cross database transforms? I know its not exposed by IIRC the code is still set up to support it (we are paying that cost?)

wotbrew commented on enterprise/backend/src/metabase_enterprise/transforms/interface.clj:
```diff
@@ -0,0 +1,31 @@
+    (transform->transform-type transform)))
+
+(defmulti dependencies
+  "Return a set of dependencies required to run this transforms."
```
Also calling putting `tables` somewhere in the name could be good

qnkhuat committed:
imporve docstring

qnkhuat committed:
remove the optiosn as it might not be applicable for all transforms

bronsa commented on enterprise/backend/src/metabase_enterprise/transforms/interface.clj:
```diff
@@ -0,0 +1,42 @@
+
+  NOTE: Currently returns a single database ID, but we may want to change this to return multiple database IDs in the
+  future since transforms can have multiple input sources."
+  {:added "0.47.0" :arglists '([transform])}
```
those should be 57 not 47

metamben reviewed (commented):
I wonder if a protocol worked better than a bunch of multimethods.

wotbrew committed:
Solve merge conflict, keep named mbql-transform function for now

wotbrew committed:
Move python specific instrumentation vars out of instrumentation.clj

wotbrew committed:
Ensure instrumentation hooks are open

Previously instrumentation ns was closed. Arguable whether better, but removes a remaining bare conditional, and removes python/mbql cases from the shared ns.

The computation stages are moved into the execution impls.
The previous placement seemed accidental? Measured against the job 'run-id' (overloaded) instead of the transforms 'run-id' as the other stages do.

wotbrew committed:
Formatting

wotbrew committed:
Fix incorrect var ref / alias

wotbrew committed:
Kondo, removed old python-runner module config

wotbrew committed:
Clarify dependency semantics

wotbrew committed:
Rename impl as suggested in review

wotbrew committed:
Add extra detail to execute! docstring

wotbrew committed:
Typo

wotbrew committed:
Fix NPE due to nil cancel-chan (merge conflict)

qnkhuat committed:
Merge branch 'master' into polymorphic-transforms

qnkhuat committed:
/s/46/57 in versioning for the multimethods

qnkhuat committed:
fix missing vars

qnkhuat committed:
Merge branch 'master' into polymorphic-transforms

qnkhuat committed:
cljfmt and kondo

qnkhuat committed:
fix kondo
